fish-rust/src/wildcard.rs

```diff
@@ -11,3 +48,1249 @@ pub const ANY_STRING_RECURSIVE: char = char_offset(WILDCARD_RESERVED_BASE, 2);
+                        // If out is NULL, we don't care about the actual matches. If out is not
+                        // NULL but we have a prefix match, stop there.
+                        let Some(out) = out.as_mut() else {
+                            return WildcardResult::Match;
+                        };
+
+                        if has_prefix_match(out, before_count) {
+                            return WildcardResult::Match;
+                        }
+                        continue;
+                    }
+                    // Note early return
+                    WildcardResult::Cancel | WildcardResult::Overflow => return submatch_res,
+                }
+            }
+
+            return match has_match {
+                true => WildcardResult::Match,
+                false => WildcardResult::NoMatch,
+            };
+        }
+        // We don't even try with this one.
+        ANY_STRING_RECURSIVE => WildcardResult::NoMatch,
+        _ => unreachable!(),
+    }
+}
+
+pub fn wildcard_complete<'f>(
+    s: &wstr,
+    wc: &wstr,
+    desc_func: Option<&'f dyn Fn(&wstr) -> WString>,
+    out: Option<&mut CompletionReceiver>,
+    expand_flags: ExpandFlags,
+    flags: CompleteFlags,
+) -> WildcardResult {
+    let params = WcCompletePack {
+        orig: s,
+        desc_func,
+        expand_flags,
+    };
+
+    return wildcard_complete_internal(s, wc, &params, flags, out, true);
+}
+
+/// Obtain a description string for the file specified by the filename.
+///
+/// The returned value is a string constant and should not be free'd.
+///
+/// \param filename The file for which to find a description string
+/// \param lstat_res The result of calling lstat on the file
+/// \param lbuf The struct buf output of calling lstat on the file
+/// \param stat_res The result of calling stat on the file
+/// \param buf The struct buf output of calling stat on the file
+/// \param err The errno value after a failed stat call on the file.
+fn file_get_desc(
+    filename: &wstr,
+    lstat: Option<fs::Metadata>,
+    stat: Option<io::Result<fs::Metadata>>,
+) -> &'static wstr {
+    let Some(lstat) = lstat else {
+        return *COMPLETE_FILE_DESC;
+    };
+
+    fn is_executable(buf: &fs::Metadata, filename: &wstr) -> bool {
+        // Weird group permissions and other such issues make it non-trivial to find out if
+        // we can actually execute a file using the result from stat. It is much safer to
+        // use the access function, since it tells us exactly what we want to know.
+        (buf.mode() as mode_t & (S_IXUSR | S_IXGRP | S_IXOTH) != 0) && waccess(filename, X_OK) == 0
+    }
+
+    // stat was only queried if lstat succeeded
+    let stat = stat.unwrap();
+    if lstat.is_symlink() {
+        return match stat {
+            Ok(stat) if stat.is_dir() => *COMPLETE_DIRECTORY_SYMLINK_DESC,
+            Ok(stat) if is_executable(&stat, filename) => *COMPLETE_EXEC_LINK_DESC,
```

Comment:
what happened to the `definitely_executable` parameter?

Reply:
Uhm, it appears I forgot to re-add that when I rebased at some point. I can fix it but that will probably lead to some minor merge conflicts in your branch

Reply:
That explains it. When you rebased, you got a conflict an resolved it in favor of your version. You could have used something like `git log riir/wildcard..origin/master src/wildcard.cpp` to show why the conflict exists.

> I can fix it but that will probably lead to some minor merge conflicts in your branch

Please do. Merge conflicts are our friends. Learn to love them.